Added SAM headers to ModelSegments CNV pipeline output.

samuelklee opened the pull request:
Comprises the commits after 7992f64.  The only commit with real substance is `Updated metadata and abstract collection classes.`.  The rest of the commits simply update calling code, related tests, and test files.  These updates were slightly less trivial for the plotting classes, so these are also split off into separate commits.

Again, probably could be engineered better (there are two parallel class hierarchies for metadata and collection classes, which is kind of gross), but we can refactor later if needed.

@asmirnov239 please review.  Again, lower priority than gCNV VCF, but the sooner this is in master the easier it will be to get things into FireCloud.  Let's try for early next week.  I'll start doc updates concurrently.

samuelklee commented on src/main/java/org/broadinstitute/hellbender/tools/copynumber/formats/collections/AbstractRecordCollection.java:
```diff
@@ -89,8 +112,11 @@ public final int size() {
     public void write(final File outputFile) {
-        try (final RecordCollectionWriter writer = new RecordCollectionWriter(outputFile)) {
-            writer.writeAllRecords(records);
+        try (final FileWriter writer = new FileWriter(outputFile, true)) {
```
This needs to be modified to first overwrite existing files with the header before appending the records.  Currently, if the file exists, this will just append everything.

samuelklee commented:
Just a note to self: now that sl_wgs_acnv is in, I did some rebasing.

My local sl_wgs_acnv_headers is rebased on master, but I haven't pushed it.
Both local and remote sl_delete_cnv_legacy are rebased on local sl_wgs_acnv_headers.
Both local and remove sl_wgs_acnv_headers_docs are rebased on top of that.
At some point sl_gcnv_ploidy_cli should be rebased as well, but currently it's rebased on pre-review sl_wgs_acnv -> sl_wgs_acnv_headers.
The forthcoming branch for adding old padding behavior to PreprocessIntervals will need to be rebased on top of sl_gcnv_ploidy_cli.

samuelklee commented on src/main/java/org/broadinstitute/hellbender/tools/copynumber/formats/collections/AbstractRecordCollection.java:
```diff
@@ -1,45 +1,56 @@
- * Represents an immutable collection of records, a set of
- * mandatory column headers given by a {@link TableColumnCollection}, and lambdas for
- * reading and writing records.
+ * Represents {@link METADATA} (which can be represented as a {@link SAMFileHeader},
```
done

samuelklee commented on src/main/java/org/broadinstitute/hellbender/tools/copynumber/formats/collections/AbstractRecordCollection.java:
```diff
@@ -1,45 +1,56 @@
-public abstract class RecordCollection<RECORD> {
+public abstract class AbstractRecordCollection<METADATA extends Metadata, RECORD> {
+    private final METADATA metadata;
+    private final SAMFileHeader header;
```
done

samuelklee commented on src/main/java/org/broadinstitute/hellbender/tools/copynumber/formats/collections/AbstractRecordCollection.java:
```diff
@@ -105,16 +131,35 @@ public boolean equals(Object o) {
     @Override
     public int hashCode() {
-        return records.hashCode();
+        int result = header.hashCode();
```
done

samuelklee commented on src/main/java/org/broadinstitute/hellbender/tools/copynumber/formats/collections/AbstractRecordCollection.java:
```diff
@@ -127,11 +172,21 @@ protected RECORD createRecord(final DataLine dataLine) {
+
+        @Override
+        protected boolean isCommentLine(final String[] line) {
+            return line.length > 0 && line[0].startsWith(COMMENT_PREFIX);
```
Yes, for now.  I think the right way to do this would be with `@PG` lines.

samuelklee commented on src/main/java/org/broadinstitute/hellbender/tools/copynumber/formats/collections/AbstractSampleRecordCollection.java:
```diff
@@ -0,0 +1,46 @@
+                                   final TableColumnCollection mandatoryColumns,
+                                   final Function<DataLine, RECORD> recordFromDataLineDecoder,
+                                   final BiConsumer<RECORD, DataLine> recordToDataLineEncoder) {
+        super(
```
done everywhere

samuelklee commented on src/main/java/org/broadinstitute/hellbender/tools/copynumber/formats/metadata/SampleLocatableMetadata.java:
```diff
@@ -0,0 +1,21 @@
+ *
+ * @author Samuel Lee &lt;slee@broadinstitute.org&gt;
+ */
+public interface SampleLocatableMetadata extends LocatableMetadata {
```
filed #3976

samuelklee commented on src/main/java/org/broadinstitute/hellbender/tools/copynumber/formats/metadata/SimpleLocatableMetadata.java:
```diff
@@ -0,0 +1,51 @@
+import org.broadinstitute.hellbender.utils.Utils;
+
+/**
+ * @author Samuel Lee &lt;slee@broadinstitute.org&gt;
```
done

samuelklee commented on src/main/java/org/broadinstitute/hellbender/tools/copynumber/models/AlleleFractionModeller.java:
```diff
@@ -69,13 +70,15 @@
         Utils.nonNull(allelicCounts);
-        Utils.nonEmpty(segments);
+        Utils.nonNull(segments);
+        Utils.validateArg(allelicCounts.getMetadata().getSequenceDictionary().equals(segments.getMetadata().getSequenceDictionary()),
```
done

samuelklee commented on src/main/java/org/broadinstitute/hellbender/tools/copynumber/plotting/PlotModeledSegments.java:
```diff
@@ -135,8 +129,19 @@ protected Object doWork() {
 
-        //load contig names and lengths from the sequence dictionary into a LinkedHashMap
-        final Map<String, Integer> contigLengthMap = PlottingUtils.getContigLengthMap(sequenceDictionaryFile, minContigLength, logger);
+        //validate sequence dictionaries and load contig names and lengths into a LinkedHashMap
```
eh, it would not be used anywhere else.

pshapiro4broad commented on src/main/java/org/broadinstitute/hellbender/tools/copynumber/caller/SimpleCopyRatioCaller.java:
```diff
@@ -0,0 +1,135 @@
+/**
+ * This caller is loosely based on the legacy ReCapSeg caller that was originally implemented in ReCapSeg v1.4.5.0,
+ * but introduces major changes.  The method is as follows:
+ * 1) use the non-log2 mean copy ratio to determine copy-neutral segments,
```
If this is going to be published as javadoc it needs to use html list elements, e.g. `<ol><li>...<li>...</ol>`. Otherwise this will all be formatted as a single paragraph, with lines wrapped together.

samuelklee replied:
Thanks @pshapiro4broad, doc updates are going to be handled in a separate PR.  Also, I think we are typically less careful about formatting javadoc that is not for CLI classes, since these will be geared more towards developers.

pshapiro4broad commented on src/main/java/org/broadinstitute/hellbender/tools/copynumber/caller/SimpleCopyRatioCaller.java:
```diff
@@ -0,0 +1,135 @@
+    }
+
+    private Statistics calculateCallingStatistics() {
+        //get the segments that fall within the copy-neutral region
```
`//` comments need a space before the first comment text char. Also IMO it's good practice to write them as complete sentences, and capitalize/punctuate them as such.

samuelklee replied:
Thanks.  We are aiming to get this PR in ASAP so I'm not going to address style issues here.

samuelklee commented on src/test/java/org/broadinstitute/hellbender/tools/copynumber/formats/collections/AbstractSampleLocatableCollectionUnitTest.java:
```diff
@@ -21,19 +23,33 @@
-    private static final String SAMPLE_NAME_EXPECTED = "test";
+
+    private static final SampleLocatableMetadata METADATA_EXPECTED = new SimpleSampleLocatableMetadata(
+            "test-sample",
```
Oops, good catch!  Added validation to CopyNumberArgumentValidationUtils.validateIntervals.

samuelklee commented on src/main/java/org/broadinstitute/hellbender/tools/copynumber/formats/collections/AbstractSampleRecordCollection.java:
```diff
@@ -0,0 +1,46 @@
+ *
+ * @author Samuel Lee &lt;slee@broadinstitute.org&gt;
+ */
+public abstract class AbstractSampleRecordCollection<RECORD> extends AbstractRecordCollection<SampleMetadata, RECORD> {
```
punting, expanded #3916.

samuelklee commented:
Thanks @asmirnov239.  Addressed comments and rebased.  @LeeTL1220 needs to make some changes on top of this branch, so please take a quick look.  Will merge once tests pass unless you catch anything major.
